gtk4/src/subclass/buildable_parser.rs

```diff
@@ -0,0 +1,185 @@
+use glib::translate::*;
+use glib::{ffi::gpointer, GString};
+use std::collections::HashMap;
+use std::mem;
+use std::ptr::NonNull;
+
+#[doc(alias = "GtkBuildableParseContext")]
+pub struct BuildableParseContext(NonNull<ffi::GtkBuildableParseContext>);
+
+impl BuildableParseContext {
+    pub(crate) fn from_raw(ctx: *mut ffi::GtkBuildableParseContext) -> Self {
+        Self(NonNull::new(ctx).expect("Null ParseContext"))
+    }
+
+    #[doc(alias = "gtk_buildable_parse_context_get_element")]
+    pub fn element(&self) -> Option<GString> {
+        unsafe {
+            from_glib_none(ffi::gtk_buildable_parse_context_get_element(
+                self.0.as_ptr(),
+            ))
+        }
+    }
+
+    #[doc(alias = "gtk_buildable_parse_context_get_element_stack")]
+    // rustdoc-stripper-ignore-next
+    /// Returns a list of tags with the last item being the currently open tag.
+    pub fn element_stack(&self) -> Vec<GString> {
+        unsafe {
+            let stack = ffi::gtk_buildable_parse_context_get_element_stack(self.0.as_ptr());
+            FromGlibPtrArrayContainerAsVec::from_glib_none_as_vec(stack)
+        }
+    }
+
+    pub fn position(&self) -> (i32, i32) {
+        unsafe {
+            let mut line_number = mem::MaybeUninit::uninit();
+            let mut char_number = mem::MaybeUninit::uninit();
+            ffi::gtk_buildable_parse_context_get_position(
+                self.0.as_ptr(),
+                line_number.as_mut_ptr(),
+                char_number.as_mut_ptr(),
+            );
+            (line_number.assume_init(), char_number.assume_init())
+        }
+    }
+}
+
+#[doc(alias = "GtkBuildableParser")]
+#[repr(C)]
+pub struct BuildableParser(NonNull<ffi::GtkBuildableParser>);
+
+impl BuildableParser {
+    pub(crate) fn from_raw(ptr: *mut ffi::GtkBuildableParser) -> Self {
+        Self(NonNull::new(ptr).expect("parser is null"))
+    }
+
+    pub(crate) fn as_ptr(&self) -> *mut ffi::GtkBuildableParser {
+        self.0.as_ptr()
+    }
+
+    pub fn new<T: BuildableParserImpl>() -> Self {
+        Self::from_raw(&mut ffi::GtkBuildableParser {
+            start_element: Some(start_element_trampoline::<T>),
+            end_element: Some(end_element_trampoline::<T>),
+            text: Some(text_trampoline::<T>),
+            error: Some(error_trampoline::<T>),
+            padding: [
+                std::ptr::null_mut(),
+                std::ptr::null_mut(),
+                std::ptr::null_mut(),
+                std::ptr::null_mut(),
+            ],
+        })
+    }
+}
+
+pub unsafe trait BuildableParserImpl: Sized {
```

Comment:
Why does it have to be unsafe?

Reply:
It shouldn't, I made it unsafe at first before figuring out what the user_data pointers are used for. Ideally once this PR is ready, we shouldn't need that